edn/src/symbols.rs

```diff
@@ -98,12 +99,73 @@ impl Keyword {
 }
 
 impl NamespacedKeyword {
+    /// Creates a new `NamespacedKeyword`. When `name` starts with a `_`, it is considered
+    /// reversed. See `symbols::NamespacedKeyword::is_reversed`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use edn::symbols::NamespacedKeyword;
+    /// let keyword = NamespacedKeyword::new("foo", "bar");
+    /// assert_eq!(keyword.to_string(), ":foo/bar");
+    /// ```
     pub fn new(namespace: &str, name: &str) -> Self {
         assert!(!name.is_empty(), "Keywords cannot be unnamed.");
         assert!(!namespace.is_empty(), "Keywords cannot have an empty non-null namespace.");
+        let reversed = match name.chars().nth(0) {
+            Some(v) => v.to_string() == "_",
+            None => false,
+        };
 
         // TODO: debug asserts to ensure that neither field matches [ :/].
-        NamespacedKeyword { name: name.to_string(), namespace: namespace.to_string() }
+        NamespacedKeyword {
+            name: if reversed { name[1..].to_string() } else { name.to_string() },
+            namespace: namespace.to_string(),
+            reversed: reversed,
+        }
+    }
+
+    /// Whether this NamespacedKeyword should be interpreted in reverse order. For example,
+    /// the two following snippets are identical:
+    /// [?y :person/friend ?x]
+    /// [?x :person/hired ?y]
+    ///
+    /// [?y :person/friend ?x]
+    /// [?y :person/_hired ?x]
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use edn::symbols::NamespacedKeyword;
+    /// assert!(!NamespacedKeyword::new("foo", "bar").is_reversed());
+    /// assert!(NamespacedKeyword::new("foo", "_bar").is_reversed());
+    /// assert_eq!(NamespacedKeyword::new("foo", "_bar").name, "bar");
+    /// ```
+    pub fn is_reversed(&self) -> bool {
+        self.reversed
+    }
+
+    /// Returns an Option<NamespacedKeyword> with the same name and namespace, but
+    /// considered reversed. See `symbols::NamespacedKeyword::is_reverse`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use edn::symbols::NamespacedKeyword;
+    /// let nsk = NamespacedKeyword::new("foo", "bar");
+    /// assert!(!nsk.is_reversed());
+    /// assert_eq!(":foo/bar", nsk.to_string());
+    ///
+    /// let reversed = nsk.to_reversed();
+    /// assert!(reversed.is_reversed());
+    /// assert_eq!(":foo/_bar", reversed.to_string());
+    /// ```
+    pub fn to_reversed(&self) -> NamespacedKeyword {
```

Comment:
I think `reverse` would imply mutability. And while I'd normally be happy with mutability in this case, the keywords we're working with will probably be borrowed references inside values.